### PR TITLE
Internal branch promotion for 2021.Q4.2

### PR DESCRIPTION
**Pull request by JaxLinAMD**

### CODE REVIEW

File: .typos.toml [hunk 1]
```diff
@@ -1,24 +0,0 @@
-# For information about the config format, see https://github.com/crate-ci/typos
```
Comment (kuhar) [inline .typos.toml:1]: Why is this file deleted?

File: include/vkgcDefs.h [hunk 2]
```diff
@@ -264,10 +264,12 @@ enum class ResourceMappingNodeType : unsigned {
...
   DescriptorConstTexelBuffer,   ///< Generic descriptor: constTexelBuffer, including uniform texel buffer
-#if (LLPC_CLIENT_INTERFACE_MAJOR_VERSION >= 50)
-  InlineBuffer, ///< Push constant with binding
+
+#if  (LLPC_CLIENT_INTERFACE_MAJOR_VERSION>= 50)
+  InlineBuffer,                 ///< Push constant with binding
 #endif
-  Count, ///< Count of resource mapping node types.
+
+  Count,                        ///< Count of resource mapping node types.
```
Comment (kuhar) [inline include/vkgcDefs.h:272]: Why is the formatting changed here?

File: lgc/interface/lgc/Pipeline.h [hunk 3]
```diff
@@ -124,7 +124,7 @@ struct Options {
   unsigned disableImageResourceCheck;  // Don't do image resource type check
   unsigned reserved0f;                 // Reserved for future functionality
   unsigned reserved10;                 // Reserved for future functionality
-  unsigned reserved1f;                 // Reserved for future functionality
+  unsigned reserved1f;         // Reserved for future functionality
```
Comment (kuhar) [inline lgc/interface/lgc/Pipeline.h:127]: Why did formatting change?

File: lgc/state/PipelineState.cpp [hunk 4]
```diff
@@ -1236,7 +1236,7 @@ void PipelineState::setShaderDefaultWaveSize(ShaderStage stage) {
       // If subgroup size is used in any shader in the pipeline, use the specified subgroup size as wave size.
       if (getShaderModes()->getAnyUseSubgroupSize()) {
         // If allowVaryWaveSize is enabled, subgroupSize is default as zero, initialized as waveSize
-        subgroupSize = getShaderOptions(checkingStage).subgroupSize;
+	subgroupSize = getShaderOptions(checkingStage).subgroupSize;
```
Comment (kuhar) [inline lgc/state/PipelineState.cpp:1239]: weird formatting

File: llpc/context/llpcCompiler.cpp [hunk 5]
```diff
@@ -1418,7 +1418,11 @@ Result Compiler::BuildGraphicsPipeline(const GraphicsPipelineBuildInfo *pipeline
   BinaryData elfBin = {};
 
   SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
-      &pipelineInfo->vs, &pipelineInfo->tcs, &pipelineInfo->tes, &pipelineInfo->gs, &pipelineInfo->fs,
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
```
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:1425]: Why did the formatting change? If you want to keep it vertical, please surround this vector with a clang-format-off/on comments

File: llpc/context/llpcGraphicsContext.cpp [hunk 6]
```diff
@@ -53,7 +53,11 @@ GraphicsContext::GraphicsContext(GfxIpVersion gfxIp, const GraphicsPipelineBuild
       m_activeStageCount(0) {
   setUnlinked(pipelineInfo->unlinked);
   const PipelineShaderInfo *shaderInfo[ShaderStageGfxCount] = {
-      &pipelineInfo->vs, &pipelineInfo->tcs, &pipelineInfo->tes, &pipelineInfo->gs, &pipelineInfo->fs,
+    &pipelineInfo->vs,
```
Comment (kuhar) [inline llpc/context/llpcGraphicsContext.cpp:56]: Also here

File: llpc/context/llpcGraphicsContext.cpp [hunk 7]
```diff
@@ -115,8 +119,12 @@ const PipelineShaderInfo *GraphicsContext::getPipelineShaderInfo(ShaderStage sha
 // =====================================================================================================================
 // Check whether the pipeline uses features relevant to subgroup size
 bool GraphicsContext::usesSubgroupSize() const {
-  std::array<const PipelineShaderInfo *, 5> shaderInfos = {
-      &m_pipelineInfo->vs, &m_pipelineInfo->tcs, &m_pipelineInfo->tes, &m_pipelineInfo->gs, &m_pipelineInfo->fs,
+  std::array<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfos = {
```
Comment (kuhar) [inline llpc/context/llpcGraphicsContext.cpp:122]: Also here

File: llpc/lower/llpcSpirvLowerGlobal.cpp [hunk 8]
```diff
@@ -1024,17 +1025,25 @@ Value *SpirvLowerGlobal::addCallInstForInOutImport(Type *inOutTy, unsigned addrS
...
-          auto elem = addCallInstForInOutImport(elemTy, addrSpace, elemMeta, elemLocOffset, maxLocOffset, elemIdx,
-                                                vertexIdx, InterpLocUnknown, nullptr, insertPos);
-          inOutValue = InsertValueInst::Create(inOutValue, elem, {idx}, "", insertPos);
+          Value *elem = nullptr;
+          if (inOutMeta.PerVertexDimension) {
+            assert(inOutMeta.InterpLoc == InterpLocCustom);
+            elem = addCallInstForInOutImport(elemTy, addrSpace, elemMeta, nullptr, 0, nullptr, 0, inOutMeta.InterpLoc,
+                                             m_builder->getInt32(idx), true, insertPos);
+          }
+          else {
```
Comment (kuhar) [inline llpc/lower/llpcSpirvLowerGlobal.cpp:1034]: This is not formatted

File: llpc/tool/amdllpc.cpp [hunk 9]
```diff
@@ -118,8 +118,8 @@ cl::opt<bool> ValidateSpirv("validate-spirv", cl::desc("Validate input SPIR-V bi
                             cl::init(true));
 
 // -entry-target: name string of entry target (for multiple entry-points)
-cl::opt<std::string> EntryTarget("entry-target", cl::desc("Name string of entry target"), cl::value_desc("entryname"),
-                                 cl::init(""));
+cl::opt<std::string> EntryTarget("entry-target", cl::desc("Name string of entry target"),cl::value_desc("entryname"),
```
Comment (kuhar) [inline llpc/tool/amdllpc.cpp:121]: Weird formatting changes here and below

File: llpc/tool/llpcAutoLayout.cpp [hunk 10]
```diff
@@ -52,6 +52,9 @@ using namespace SPIRV;
 namespace Llpc {
 namespace StandaloneCompiler {
 
+// Offset in Node will be a fixed number, which will be conveniently to identify offset in auto-layout.
+static const unsigned OffsetStrideInDwords = 12;
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:56]: nit: `constexpr unsigned OffsetStrideInDwords = 12;`

File: llpc/tool/llpcAutoLayout.cpp [hunk 11]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+
+// =====================================================================================================================
+// Find userDataNode with specified set and binding. And return Node index.
+//
+// @param userDataNode : ResourceMappingNode pointer
+// @param nodeCount : User data node count
+// @param set : Find same set in node array
+// @param binding : Find same binding in node array
+// @param [out] index : Return node position in node array
+static const ResourceMappingNode *findResourceNode(const ResourceMappingNode *userDataNode, unsigned nodeCount,
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:115]: `@returns` is missing

File: llpc/tool/llpcAutoLayout.cpp [hunk 12]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+
+  return resourceNode;
+}
+
+// =====================================================================================================================
+// Check if autoLayoutUserDataNodes is a subset of userDataNodes.
+//
+// @param [in] resourceMapping : Resource mapping data, which can contain user data nodes
+// @param [in] autoLayoutUserDataNodeCount : UserData Node count
+// @param [in] autoLayoutUserDataNodes : ResourceMappingNode
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:163]: No need for `[in]`

File: llpc/tool/llpcAutoLayout.cpp [hunk 13]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+                       << format("0x%016" PRIX64, static_cast<unsigned>(autoLayoutNext->type))
+                       << "\n ->sizeInDwords            : " << autoLayoutNext->sizeInDwords
+                       << "\n ->offsetInDwords          : " << autoLayoutNext->offsetInDwords;
+
+                outs() << "\nShaderInfoNode:"
+                       << "\n ->type                    : "
+                       << format("0x%016" PRIX64, static_cast<unsigned>(node->type))
+                       << "\n ->sizeInDwords            : " << node->sizeInDwords
+                       << "\n OffsetStrideInDwords      : " << OffsetStrideInDwords
+                       << "\n index*OffsetStrideInDwords: " << (index * OffsetStrideInDwords) << "\n";
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:210]: Shouldn't this use LLPC_OUTS?

File: llpc/tool/llpcAutoLayout.cpp [hunk 14]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+// @param [in] resourceMapping : Resource mapping data, which can contain user data nodes
+// @param [in] autoLayoutUserDataNodeCount : UserData Node count
+// @param [in] autoLayoutUserDataNodes : ResourceMappingNode
+// @returns : true if compatible
+bool checkResourceMappingComptible(const ResourceMappingData *resourceMapping, unsigned autoLayoutUserDataNodeCount,
+                                   const ResourceMappingRootNode *autoLayoutUserDataNodes) {
+  bool hit = false;
+
+  if (autoLayoutUserDataNodeCount == 0)
+    hit = true;
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:170]: Use early returns

File: llpc/tool/llpcAutoLayout.cpp [hunk 15]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+  else if (resourceMapping->pStaticDescriptorValues)
+    hit = false;
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:172]: Use an early return

File: llpc/tool/llpcAutoLayout.cpp [hunk 16]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+}
+
+// =====================================================================================================================
+// Check if autoLayoutUserDataNodes is a subset of userDataNodes.
+//
+// @param [in] resourceMapping : Resource mapping data, which can contain user data nodes
+// @param [in] autoLayoutUserDataNodeCount : UserData Node count
+// @param [in] autoLayoutUserDataNodes : ResourceMappingNode
+// @returns : true if compatible
+bool checkResourceMappingComptible(const ResourceMappingData *resourceMapping, unsigned autoLayoutUserDataNodeCount,
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:165]: This function is very long. It should be split into smaller helper functions to make it more maintainable. It also deserved unit tests.

File: llpc/tool/llpcAutoLayout.cpp [hunk 17]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+// @returns : true if compatible
+bool checkPipelineStateCompatible(const ICompiler *compiler, Llpc::GraphicsPipelineBuildInfo *pipelineInfo,
+                                  Llpc::GraphicsPipelineBuildInfo *autoLayoutPipelineInfo, Llpc::GfxIpVersion gfxIp) {
+  bool compatible = true;
+
+  auto cbState = &pipelineInfo->cbState;
+  auto autoLayoutCbState = &autoLayoutPipelineInfo->cbState;
+
+  for (unsigned i = 0; i < MaxColorTargets; ++i) {
+    if (cbState->target[i].format != VK_FORMAT_UNDEFINED) {
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:263]: Use continue if the format is undefined

File: llpc/tool/llpcAutoLayout.cpp [hunk 18]
```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
...
+      // NOTE: Alpha-to-coverage only take effect for output from color target 0.
+      const bool enableAlphaToCoverage = cbState->alphaToCoverageEnable && i == 0;
+      unsigned exportFormat =
+          compiler->ConvertColorBufferFormatToExportFormat(&cbState->target[i], enableAlphaToCoverage);
+      const bool autoLayoutEnableAlphaToCoverage = autoLayoutCbState->alphaToCoverageEnable && i == 0;
+      unsigned autoLayoutExportFormat = compiler->ConvertColorBufferFormatToExportFormat(
+          &autoLayoutCbState->target[i], autoLayoutEnableAlphaToCoverage);
+
+      if (exportFormat != autoLayoutExportFormat) {
+        outs() << "pPipelineInfo->cbState.target[" << i << "] export format:" << format("0x%016" PRIX64, exportFormat)
```
Comment (kuhar) [inline llpc/tool/llpcAutoLayout.cpp:273]: LLPC_OUTS?

File: llpc/tool/llpcCompilationUtils.cpp [hunk 19]
```diff
@@ -475,7 +475,6 @@ Error processInputStages(ICompiler *compiler, CompileInfo &compileInfo, ArrayRef
             return createResultError(Result::ErrorInvalidShader, Twine("Failed to validate SPIR-V:\n") + log);
         }
       }
-
```
Comment (kuhar) [inline llpc/tool/llpcCompilationUtils.cpp:478]: Weird formatting change

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 20]
```diff
@@ -1471,7 +1471,8 @@ Value *SPIRVToLLVM::addLoadInstRecursively(SPIRVType *const spvType, Value *load
 
   Constant *const zero = getBuilder()->getInt32(0);
 
-  if (loadType->isStructTy() && spvType->getOpCode() != OpTypeSampledImage && spvType->getOpCode() != OpTypeImage) {
+  if (loadType->isStructTy() && spvType->getOpCode() != OpTypeSampledImage && spvType->getOpCode() != OpTypeImage
+  ) {
```
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:1475]: Weird formatting changes here and below

File: llpc/unittests/standaloneCompiler/testInputUtils.cpp [hunk 21]
```diff
@@ -205,6 +205,41 @@ TEST(InputUtilsTest, IsSpirvBinaryFile) {
 
 TEST(InputUtilsTest, IsGlslShaderFile) {
   // Based on https://www.khronos.org/opengles/sdk/tools/Reference-Compiler/.
+  //
+  //Copyright (C) 2002-2005 3Dlabs Inc. Ltd.
+  //Copyright (C) 2012-2013 LunarG, Inc.
```
Comment (kuhar) [inline llpc/unittests/standaloneCompiler/testInputUtils.cpp:210]: This is not necessary IMO. The only thing based on the website is the list of extensions. The implementation is not based on the website.

File: llpc/util/llpcDebug.h [hunk 22]
```diff
@@ -52,7 +52,6 @@
 namespace MetroHash {
 struct Hash;
 } // namespace MetroHash
-
```
Comment (kuhar) [inline llpc/util/llpcDebug.h:55]: weird formatting change

File: tool/dumper/vkgcPipelineDumper.cpp [hunk 23]
```diff
@@ -431,7 +431,7 @@ void PipelineDumper::dumpResourceMappingNode(const ResourceMappingNode *userData
   case ResourceMappingNodeType::DescriptorFmask:
   case ResourceMappingNodeType::DescriptorBufferCompact:
   case ResourceMappingNodeType::PushConst:
-#if (LLPC_CLIENT_INTERFACE_MAJOR_VERSION >= 50)
+#if  (LLPC_CLIENT_INTERFACE_MAJOR_VERSION>= 50)
```
Comment (kuhar) [inline tool/dumper/vkgcPipelineDumper.cpp:434]: formatting

File: tool/dumper/vkgcPipelineDumper.cpp [hunk 24]
```diff
@@ -813,7 +813,11 @@ void PipelineDumper::dumpGraphicsPipelineInfo(std::ostream *dumpFile, const char
   dumpVersionInfo(*dumpFile);
   // Dump pipeline
   const PipelineShaderInfo *shaderInfos[ShaderStageGfxCount] = {
-      &pipelineInfo->vs, &pipelineInfo->tcs, &pipelineInfo->tes, &pipelineInfo->gs, &pipelineInfo->fs,
+    &pipelineInfo->vs,
```
Comment (kuhar) [inline tool/dumper/vkgcPipelineDumper.cpp:816]: Also here

File: tool/dumper/vkgcPipelineDumper.cpp [hunk 25]
```diff
@@ -831,7 +835,7 @@ void PipelineDumper::dumpGraphicsPipelineInfo(std::ostream *dumpFile, const char
 }
 
 // =====================================================================================================================
-// Builds hash code from graphics pipeline build info.  If stage is a specific stage of the graphics pipeline, then only
+// Builds hash code from graphics pipline build info.  If stage is a specific stage of the graphics pipeline, then only
```
Comment (kuhar) [inline tool/dumper/vkgcPipelineDumper.cpp:838]: this introduces a typo

File: tool/dumper/vkgcPipelineDumper.cpp [hunk 26]
```diff
@@ -1237,7 +1241,7 @@ void PipelineDumper::updateHashForResourceMappingNode(const ResourceMappingNode
     // Do nothing for the stream-out table
     break;
   }
-#if (LLPC_CLIENT_INTERFACE_MAJOR_VERSION >= 50)
+#if  (LLPC_CLIENT_INTERFACE_MAJOR_VERSION>= 50)
```
Comment (kuhar) [inline tool/dumper/vkgcPipelineDumper.cpp:1244]: formatting

File: tool/vfx/vfxPipelineDoc.cpp [hunk 27]
```diff
@@ -152,8 +152,12 @@ VfxPipelineStatePtr PipelineDocument::getDocument() {
   if (m_pipelineState.pipelineType == VfxPipelineTypeGraphics ||
       m_pipelineState.pipelineType == VfxPipelineTypeCompute) {
     PipelineShaderInfo *shaderInfo[NativeShaderStageCount] = {
-        &m_pipelineState.gfxPipelineInfo.vs, &m_pipelineState.gfxPipelineInfo.tcs, &m_pipelineState.gfxPipelineInfo.tes,
-        &m_pipelineState.gfxPipelineInfo.gs, &m_pipelineState.gfxPipelineInfo.fs,  &m_pipelineState.compPipelineInfo.cs,
+      &m_pipelineState.gfxPipelineInfo.vs,
```
Comment (kuhar) [inline tool/vfx/vfxPipelineDoc.cpp:155]: also here

**Review by kuhar — CHANGES_REQUESTED** [review]
This import seems to have a lot of accidental edits and weird formatting changes. Please undo all accidental changes and surround the code with `// clang-format-off` /  `// clang-format-on` if you want it to be formatted differently.

Also, the new code in autolayout seems untested. If this needs to be promoted quickly, either exclude it (seems unused at the moment), or we should have a post-merge review later.

File: llpc/context/llpcCompiler.cpp (hunk 5 above)
Comment (JaxLinAMD) [reply llpc/context/llpcCompiler.cpp:1425]: internal code is the format of:
&pipelineInfo->vs,
#if XXXX
some code here
&pipelineInfo->tcs,

so need to keep it vertical.

**Comment by JaxLinAMD** [comment]
@kuhar the formatting change is caused by sanitization script, we need to fix the bug of the internal script. after the script is fixed, there will no be such format changes. Can we let the format changes be there until we fix the script?

File: llpc/context/llpcCompiler.cpp (hunk 5 above)
Comment (kuhar) [reply llpc/context/llpcCompiler.cpp:1425]: Use clang-format on / off: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code.
This PR should pass the code formatting check

**Comment by kuhar** [comment]
> @kuhar the formatting change is caused by sanitization script, we need to fix the bug of the internal script. after the script is fixed, there will no be such format changes. Can we let the format changes be there until we fix the script?

Can you run clang-format after the sanitization script? It seems difficult for me to to make sanitization and formatting work as one step. If you want some pieces of code to be formatted differently (e.g., the arrays with stages being vertical), surround such code with `// clang-format on` / `off`. 

The issue with leaving this as-is that someone has to to a follow-up clean up, which may end up being a very low priority task and happen after a long time, when it's difficult to track down all the little places that need to be fixed. And that would have to go through another code review, etc. This only doubles the effort. From my point of view, making the PR correct the first time seems like something that is the least time-consuming option.

**Comment by kuhar** [comment]
@JaxLinAMD or alternatively, if you want this to be the synchronization point for internal codebase and upstream, you can leave the commit from the PR as-is and add a second commit that does the cleanup. This way the current commit stays as-is for the reasons you mentioned, and we keep the state of the upstream clean because we submit a cleanup commit at the same time. How does this sound?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0a213b235abeeee67491bdeeccd1a5a1cc90c45c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev f49a7f3f706dd721d9a7b620be1482217ef63b6e</li>
<li>SPVGEN: dev efa3cec20c6ebc66254ff9cca9b3ea124811d7a3</li>
<li>XGL: dev c3d047085243c811583a612b5ae016a513705465</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cdd74831d061bf65fdf5af4a0b6ce99fe2c558a9</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@kuhar hmm, yes, you're right. Let me just fix the format in this PR.
